pkg/volume/csi/csi_plugin.go

```diff
@@ -366,23 +406,61 @@ func (p *csiPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.S
 
 	klog.V(4).Info(log("plugin.ConstructVolumeSpec extracted [%#v]", volData))
 
-	fsMode := api.PersistentVolumeFilesystem
-	pv := &api.PersistentVolume{
-		ObjectMeta: meta.ObjectMeta{
-			Name: volData[volDataKey.specVolID],
-		},
-		Spec: api.PersistentVolumeSpec{
-			PersistentVolumeSource: api.PersistentVolumeSource{
-				CSI: &api.CSIPersistentVolumeSource{
-					Driver:       volData[volDataKey.driverName],
-					VolumeHandle: volData[volDataKey.volHandle],
+	var spec *volume.Spec
+	inlineEnabled := utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume)
+
+	if inlineEnabled {
+		mode := driverMode(volData[volDataKey.driverMode])
+		switch {
+		case mode == ephemeralDriverMode:
+			vol := &api.Volume{
+				Name: volData[volDataKey.specVolID],
+				VolumeSource: api.VolumeSource{
+					CSI: &api.CSIVolumeSource{
+						Driver: volData[volDataKey.driverName],
+					},
+				},
+			}
+			spec = volume.NewSpecFromVolume(vol)
+
+		case mode == persistentDriverMode:
+			fsMode := api.PersistentVolumeFilesystem
```

Comment:
this block is the same as the one below, can we consolidate them?

Reply:
yes fixed. also bad comments fixed.